forc-pkg/src/manifest.rs

```diff
@@ -549,30 +553,52 @@ impl WorkspaceManifestFile {
         Self::from_file(path)
     }
 
+    /// Return an iterator over the names of declared workspace members
     pub fn members(&self) -> impl Iterator<Item = &String> + '_ {
         self.members.iter()
     }
 
+    /// Return an iterator over the paths of declared workspace members
     pub fn member_paths(&self) -> Result<impl Iterator<Item = PathBuf> + '_> {
         let parent = self
             .path
             .parent()
             .ok_or_else(|| anyhow!("Cannot get parent dir of {:?}", self.path))?;
         Ok(self.members.iter().map(|member| parent.join(member)))
     }
+
+    /// The path to the `Forc.toml` from which this manifest was loaded.
+    ///
+    /// This will always be a canonical path.
+    pub fn path(&self) -> &Path {
+        &self.path
+    }
+
+    /// The path to the directory containing the `Forc.toml` from which this manfiest was loaded.
+    ///
+    /// This will always be a canonical path.
+    pub fn dir(&self) -> &Path {
+        self.path()
+            .parent()
+            .expect("failed to retrieve manifest directory")
+    }
 }
 
 impl WorkspaceManifest {
     /// Given a path to a `Forc.toml`, read it and construct a `WorkspaceManifest`.
     pub fn from_file(path: &Path) -> Result<Self> {
+        let mut warnings = Vec::new();
         let manifest_str = std::fs::read_to_string(path)
             .map_err(|e| anyhow!("failed to read manifest at {:?}: {}", path, e))?;
         let toml_de = &mut toml::de::Deserializer::new(&manifest_str);
         let manifest: Self = serde_ignored::deserialize(toml_de, |path| {
             let warning = format!("  WARNING! unused manifest key: {}", path);
-            println_yellow_err(&warning);
+            warnings.push(warning);
```

Comment:
I had to hack this like that because to understand whether we have a package manifest or a workspace manifest, I am trying to create a `PackageManifestFile` from the dir I am checking. If that fails I am trying to see if current manifest is a `WorkspaceManifestFile`. If the manifest I am checking is indeed a `WorkspaceManifest`, checking to see if it was a `PackageManifestFile` prints out warnings of `unused key: members` as unused warnings are printed before checking success of parsing.